src/coreclr/nativeaot/Runtime/stdcppshim.cpp

```diff
@@ -0,0 +1,24 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "CommonTypes.h"
+
+void* operator new(size_t n, const nothrow_t&) noexcept
+{
+    return malloc(n);
+}
+
+void* operator new[](size_t n, const nothrow_t&) noexcept
+{
+    return malloc(n);
+}
+
+void operator delete(void *p) noexcept
```

Comment:
Can we do something about these so that they are not globally visible? Make them static?

Reply:
They are not visible outside the object. It is the same (global scoped) approach in `src/coreclr/utilcode/clrhost_nodependencies.cpp`.

The standard doesn't allow namespace scoped operator delete, it allows class/struct scoped one, but they cannot be used in global scope.

Reply:
What makes it not visible outside the object? What would an implementation that is visible outside the object look like?

> It is the same (global scoped) approach in src/coreclr/utilcode/clrhost_nodependencies.cpp.

I know we had been fighting problems with this one, and I do not think we found good solution

Reply:
With `extern "C"`, it will have a visible symbol which will conflict with other object depending on libstdc++.

Reply:
The symbol should still be visible though. There is nothing here that stops them from being visible during link. I could be missing something here, but they are defined in the header and not marked `static` so they can be discovered outside of the compilation unit. Is this some NativeAOT pattern I'm missing?

Reply:
> There is nothing here that stops them from being visible during link. 

There is this: https://github.com/dotnet/runtime/blob/e28cbaed0fdda578d7e365495e7821bfeb58e194/eng/native/configurecompiler.cmake#L482

Reply:
This controls visibility in the .so after linking. I do not think that this controls visibility between multiple .o files.

Reply:
Perhaps we aren't talking about the same thing or I could be misunderstanding something with that flag. My reading of `-fvisibility` is it relates to post-link exports of a dynamic library (`.so`) not the object files creating during compilation and then assembled at link-time. That is at least how `__declspec(dllexport)` works. It doesn't hide them from other compilation units during link, rather it presents them as exports from the linked library (that is, `.dll`). Those functions will still be visible to other compilation units pre-link.

Reply:
I image there is no tests for the scenario we are worried will break, right? Is it important enough to block this PR on?

Reply:
> I image there is no tests for the scenario we are worried will break, right? 

I would agree with that. @jkotas's concern is something I share and along with @janvorli have had to chase down - binding to incorrect allocators. This is likely far more manageable in this narrow slice of the world, but it has caused us a non-trivial amount of stress and fire drills, particularly in diagnostic scenarios.

Reply:
One option is that we create a [`gcompat`](https://git.adelielinux.org/adelie/gcompat/) type object with these few definitions and link it by default. For static NativeAOT publishing, we guide the user to drop it if they are using `-lstdc++`.

Reply:
I would still keep most of the changes you have made in this PR, so that the compat lib can be minimal.